### PR TITLE
Prepare for want patterns

### PR DESCRIPTION
**Pull request by erights**

### CODE REVIEW

File: packages/zoe/src/zoeService/types.js [hunk 1]
```diff
@@ -271,7 +271,7 @@
  */
 
 /**
- * @typedef {Partial<ProposalRecord>} Proposal
+ * @typedef {Partial<ProposalRecord>}   Proposal
```
Comment (erights) [inline packages/zoe/src/zoeService/types.js]: bad whitespace

Comment (FUDCo) [reply packages/zoe/src/zoeService/types.js]: not bad, just misunderstood

**Review by FUDCo — APPROVED** [review]
This seems fine to me, though might also want to run it past somebody with more domain-specific paranoia.

**Comment by erights** [comment]
Hi @Chris-Hibbert , we need your eyeballs again ;). Thanks.

File: packages/zoe/src/instanceRecordStorage.js [hunk 2]
```diff
@@ -75,7 +77,7 @@ export const makeInstanceRecordStorage = () => {
     assertInstantiated();
     assertKeywordName(keyword);
     assert(
-      !getKeywords(instanceRecord.terms.issuers).includes(keyword),
+      !ownKeys(instanceRecord.terms.issuers).includes(keyword),
```
Comment (Chris-Hibbert) [inline packages/zoe/src/instanceRecordStorage.js:80]: is `Reflect.ownKeys()` equivalent to `Object.getOwnPropertyNames()`?

I'd sure prefer to use  the former if I can. If it's substitutable, I'll do the search and replace myself. We use the latter a lot, and I find it really ugly and verbose.

File: packages/zoe/test/unitTests/test-cleanProposal.js [hunk 3]
```diff
@@ -130,15 +145,60 @@ test('cleanProposal - other wrong stuff', t => {
...
+    'nat',
+    /"foo" - Must have passStyle or tag "remotable"/,
+  );
+  proposeBad(
+    { exit: { afterDeadline: { timer, deadline: 'foo' } } },
+    'nat',
+    /"foo" - Must be >= "\[0n\]"/,
+  );
+  proposeBad(
+    { exit: { afterDeadline: { timer: 'foo', deadline: 3n } } },
```
Comment (Chris-Hibbert) [inline packages/zoe/test/unitTests/test-cleanProposal.js]: redundant with line 163

File: packages/zoe/src/cleanProposal.js [hunk 4]
```diff
@@ -40,116 +37,83 @@ export const assertKeywordName = keyword => {
...
+      // TODO: replace this assertion with a check of the assetKind
+      // property on the brand, when that exists.
+      assert(
+        assetKind === brandAssetKind,
+        X`The amount ${amount} did not have the assetKind of the brand ${brandAssetKind}`,
+      );
+      return AmountMath.coerce(amount.brand, amount);
+    } else {
+      assertPattern(amount);
+      return amount;
```
Comment (Chris-Hibbert) [inline packages/zoe/src/cleanProposal.js:78]: I don't think there are any tests of this path.  Should it be included in this PR?

**Review by Chris-Hibbert — COMMENTED** [review]
This looks good to me modulo
 * I don't have enough context to evaluate the changes to `patternMathers.js`
 * `coerceAmountPatternKeywordRecord()` allows amounts that are patterns, but there don't seem to be tests of this path. Is this premature?

File: packages/zoe/src/zoeService/types.js (hunk 1 above)
Comment (erights) [reply packages/zoe/src/zoeService/types.js]: Seeing this, I immediately lept to "I'm not bad, I'm just drawn that way"

image

Done

File: packages/zoe/src/instanceRecordStorage.js (hunk 2 above)
Comment (erights) [reply packages/zoe/src/instanceRecordStorage.js:80]: Once something is known to be a copyArray, all the following are guaranteed to enumerate exactly the same set of property names:
   * `Object.getOwnPropertyNames` (string-named own property names)
   * `Object.keys` (string named enumerable own property names)
   * `Object.assign` (enumerable own property names)
   * `Reflect.ownKeys` (all own property names)
   * `...` (enumerable own property names)
   * `for/in` loop (enumerable property names, including inherited ones)

The reason is that a copyRecord 
   * must directly inherit either from nothing or from `Object.prototype`, which is hardened and has no enumerable properties.
   * All own properties are string-named and enumerable.

copyArrays are similar, except that the magic array `.length` property is non-enumerable, and so will be enumerated only by `getOwnPropertyNames` and `ownKeys` from the above list.

For all objects not known to be copyRecords or copyArrays, no such assumption should be made.

Comment (erights) [reply packages/zoe/src/instanceRecordStorage.js:80]: No doubt there are many other occurrences in our code that would be improved by switching to `ownKeys`.

Comment (Chris-Hibbert) [reply packages/zoe/src/instanceRecordStorage.js:80]: > copyArrays are similar, except that the magic array .length property is non-enumerable, and so will be enumerated only by getOwnPropertyNames and ownKeys from the above list.

That sounds like `getOwnPropertyNames` will sometimes produce `length` as one of its results, which makes me leery of using `getOwnPropertyNames` as the general solution for enumerating field names.  Did I read that wrong?

File: packages/zoe/src/cleanProposal.js (hunk 4 above)
Comment (erights) [reply packages/zoe/src/cleanProposal.js:78]: Thanks for catching this oversight. I improved the cleanProposal tests a lot, including for want patterns.

**Pull request merged**